### PR TITLE
Add initial cookie sharing samples

### PR DESCRIPTION
**Pull request by HaoK**
Initial PR merging over the two Core CookieSamples (With/out identity) from @guardrex  PR here: https://github.com/aspnet/AuthSamples/pull/24

I will submit a follow up PR adding functional tests + doing renaming/cleanup

The old precore interop sample will come in a different PR as well

### CODE REVIEW

**Comment by HaoK** [comment]
I'm still not thrilled with the current names, namely I think CookieSharing.Core is not clear at all what that is.

How about something like:

```
CookieSharing.OnlyCookies
CookieSharing.IdentityCore
CookieSharing.IdentityNetFramework
```

**Comment by guardrex** [comment]
It was done to distinguish ...

| App | Description |
| --- | --- |
| CookieSharing.Core                      | ASP.NET Core 2.0 Razor Pages app without using ASP.NET Core Identity |
| CookieSharing.WithIdentity.Core         | ASP.NET Core 2.0 MVC app with ASP.NET Core Identity |
| CookieSharing.WithIdentity.NETFramework | ASP.NET Framework 4.6.1 MVC app with ASP.NET Identity |

When you show "OnlyCookies," it's not clear that it's a Core app ... it could be an only-cookies netfx app.

Then, "PreCore" is unknown to devs ... I would guess anyway.

**Comment by guardrex** [comment]
```
CookieSharing.OnlyCookiesCore
CookieSharing.IdentityCore
CookieSharing.IdentityNETFramework --or-- CookieSharing.IdentityNetFramework
```
?

**Comment by HaoK** [comment]
Yeah PreCore is crappy for sure :) OldIdentity?  The reason I don't like IdentityNetFramework is that you can target net461 in core as well...

**Comment by HaoK** [comment]
Or we could just use identity as the axis.
```
Sharing.NoIdentity
Sharing.IdentityCore
Sharing.IdentityNonCore
```

?

**Comment by HaoK** [comment]
Cookies are involved by default given that this is under CookieSharing, so really its just core / identity that are the moving bits

File: samples/CookieSharing/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Cookie Sharing Sample App
+
+The sample illustrates cookie sharing across three apps that use cookie authentication:
+
+| Project                                 | Description |
+| --------------------------------------- | ----------- |
+| CookieSharing.Core                      | ASP.NET Core 2.0 Razor Pages app without using ASP.NET Core Identity |
+| CookieSharing.WithIdentity.Core         | ASP.NET Core 2.0 MVC app with ASP.NET Core Identity |
+| CookieSharing.WithIdentity.NETFramework | ASP.NET Framework 4.6.1 MVC app with ASP.NET Identity |
```
Comment (Tratcher) [inline samples/CookieSharing/README.md]: I understand sharing a login between WithIdentity.Core and WithIdentity.NETFramework, but why do you have to register twice? Wouldn't you also want to share the user database between the two?

Also, sharing a cookie between identity and non-identity apps doesn't make much sense. That Core non-Identity sample should be sharing with a .NET non-Identity sample.

**Comment by guardrex** [comment]
> you can target net461 in core as well

I see what you mean, and that's always been a pain point in describing app frameworks.

I think its clear SxS with "Core" in the other names tho ... maybe! lol ... that it's a netfx app when it says any of "NETFX," "NETFramework," or "NetFramework."

Whether it's "NonCore" or "PreCore," those would be brand new terms/designations AFAICT. Is it best to ask Dan what he wants to do here?

**Comment by HaoK** [comment]
Sure, @danroth27 @blowdart @ajcvickers can you guys help us name our cookie compat sample apps?

File: samples/CookieSharing/README.md (hunk 1 above)
Comment (HaoK) [reply samples/CookieSharing/README.md]: I think each of these topics is complicated enough that its reasonable to split them up into Cookie Sharing, and in the future if we really need the shared identity database with old + new mvc identity sample

Comment (Tratcher) [reply samples/CookieSharing/README.md]: Then why are we providing this sample? Sharing the cookies doesn't seem that useful if the information behind them isn't the same.

Comment (HaoK) [reply samples/CookieSharing/README.md]: We've definitely gotten asks about how to share cookies before, and its also a required part of any kind of interop between old/new apps.  This part is the same excluding identity, interop with identity is just a more advanced variation that requires additional stuff to make identity talk to the same db.

Comment (HaoK) [reply samples/CookieSharing/README.md]: This sample is focused on sharing the same auth cookie across apps, but not focusing on how you create the cookie/claims.

Comment (Tratcher) [reply samples/CookieSharing/README.md]: It doesn't make a lot of sense in isolation. Maybe I'm too far removed from the customer requests. Have any backreferences? #34 doesn't have any.

Comment (HaoK) [reply samples/CookieSharing/README.md]: I don't remember the issues off the top of my head, but specifically these samples are for this doc: https://docs.microsoft.com/en-us/aspnet/core/security/cookie-sharing?view=aspnetcore-2.1&tabs=aspnetcore2x

Comment (guardrex) [reply samples/CookieSharing/README.md]: I plan to remove the samples from the docs repo and perform cross-repo code block sample links. 

How many times have I done this?

sulu-first-attempt

*Actually, it's my first attempt.* - Mr. Sulu ([George Takei](https://georgetakei.com/))
[Star Trek V: The Final Frontier](http://www.startrek.com/), ©1989 [Paramount Pictures](http://www.paramount.com/movies/star-trek-v-final-frontier)

**Comment by HaoK** [comment]
Ok that's a decent compromise, the Core samples will be part of the build/have tests, and the pre core one will just be there mostly as sample code but won't be runnable like the core ones

**Comment by guardrex** [comment]
Note the alternative to this PR is for us (docs repo) to keep our current samples. If this closes, then I'll close https://github.com/aspnet/Docs/issues/5238.

**Comment by HaoK** [comment]
How important is this to you @guardrex are you ok with leaving this all in the docs as is?

**Comment by HaoK** [comment]
I think the only benefit of this as an auth samples is we would have some functional verification of sharing cookies which was the main driver.  @Tratcher did we kill the interop tests for cookies or do those still exist?

**Comment by Tratcher** [comment]
They're still [around](https://github.com/aspnet/Security/blob/master/test/Microsoft.Owin.Security.Interop.Test/CookieInteropTests.cs), but there's not much coverage there.

**Comment by HaoK** [comment]
That's good enuf I think, the additional coverage we'd get from the auth samples is basically just a different way to wire up the data protection, at the end of the day we are just trying to test cookie interop which those tests seem to already cover

**Comment by guardrex** [comment]
> I think the only benefit ...

You said earlier that you wanted devs to see how to use a source-of-truth for the identity ... a common Identity dB. Our doc sample doesn't exhibit that scenario. IIRC, we have a collection of in-memory dBs (as we normally use for doc samples) ... so users have to be registered in each one.

I am good if you want to leave the dB particulars to the dev to work out. I can keep up with other cookie sharing updates. I don't think we expect major changes release-to-release. I find out pretty fast, too, when I go to upgrade the sample each time ..... if it :boom: on me, I ping you ..... if it :smile:, we're good.

**Comment by guardrex** [comment]
:point_up: ... the other possibility is that you **_don't** proceed here with a common Identity dB but we add a section to the topic (text only) to outline the basics of how that works ..... or ... or ... we do it ... but we do it over in the docs sample.

**Comment by HaoK** [comment]
The main reason I don't think its worth it anymore is we are a long ways away from pre core identity and I haven't seen any traffic asking about that kind of interop, which makes me think its not that important for us to have a real sample for this until we actually get requests about this...  Adding some text to the docs seems good enuf to me

Thoughts @ajcvickers @blowdart ?

**Comment by blowdart** [comment]
It was always meant to be a temporary solution until you moved all your apps up, or moved to a federated identity. After 2+ years if you still haven't done that, well, honestly I don't know how much more help we can give. I have it in a repo of my own, I don't care either way here.

**Comment by HaoK** [comment]
Okay, so then I think adding this to auth samples isn't worth it...  This scenario purely becomes something in the docs, and we can always add more tests to the cookie interop tests if we ever start regressing things

@guardrex do you need any help from us with respect to the doc updates you suggested?

**Comment by guardrex** [comment]
> Adding some text to the docs seems good enuf to me

> don't know how much more help we can give

I'll leave that docs issue open then. I need a few notes that I can wordsmith into the doc text. What are the minimum considerations for using a common dB across the apps? You don't need to tell me now .... I'm (and you're probably, too) 🏃🏃🏃🏃🏃🏃 at the moment. Add notes to the https://github.com/aspnet/Docs/issues/5238 issue. I'll get to it as soon as I can.

**Comment by guardrex** [comment]
> need any help

Let's just start with a bullet list of the main points. I'll see what I can do with that. We can discuss the possibility of making the docs sample use a common dB over on that issue when I get to it.

**Pull request closed without merging**